### ISSUE TITLE
[ENHANCEMENT] Support for FusedLinear+CrossEntropy Loss

### ISSUE THREAD

**Issue by Skylion007** [NVIDIA/Megatron-LM#1738]
**Is your feature request related to a problem? Please describe.**
We already support FusedCrossEntropy losses, it would be even better to support fusing the last linear layer and the CrossEntropy layer to ensure even more memory savings. See https://github.com/pytorch/pytorch/issues/124480 for more discussion. There is talk of adding this as a native operator in PyTorch to help facilitate this. Maybe this issue would be better open in TransformerEngine though. This would massively reduce the amount of memory needed for finetuning small models with large vocabularies or large vocabularies period.

**Describe the solution you'd like**
Support fusing the final linear layer with the cross entropy layer in a dtype aware way that avoid unnecessary computation based the precision of the floating point.

**Describe alternatives you've considered**

**Proposed implementation**
If you have a proposed implementation for the feature state it here or link to a PR.

See the linked Liger kernel or Apple's Cut Cross Entropy kernel.

**Additional context**

**Comment by sbhavani** [comment]
@Skylion007  thanks for sharing this feature! It'd be great if you had any perf comparisons with this feature. 

And I agree this might be a better fit for Transformer Engine but Megatron is still a good place to prove the memory savings benefits.

**Comment by Skylion007** [comment]
Figure 1 in this paper demonstrates how useful this feature can be when finetuning LLMs. Table 1 provides benchmark numbers https://arxiv.org/abs/2411.09009

**Comment by sbhavani** [comment]
@ksivaman

**Comment by Pedrexus** [comment]
Any update on this? 

I am currently trying to compare between the regular Megatron loss, the Liger impl and the CCE (Apple) impl. It seems they do save quite a considerable amount of memory in some cases, but apparently there are some numerical errors that accumulate over time.
